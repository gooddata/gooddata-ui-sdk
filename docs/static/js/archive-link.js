document.addEventListener('DOMContentLoaded', function () {
    function createArchiveLink() {
        const link = document.createElement('a');
        link.className = 'dropdown-item';
        
        const currentHostname = window.location.hostname;
        if (currentHostname.includes('netlify.app')) {
            // For Netlify previews, link to the versions page within the current preview site
            link.href = window.location.origin + '/docs/gooddata-ui/versions/';
        } else if (currentHostname === 'localhost' || currentHostname === '127.0.0.1') {
            // For local testing, link to the versions page
            link.href = '/versions/';
        } else {
            // For production - link to the unversioned archive page
            link.href = 'https://www.gooddata.com/docs/gooddata-ui/versions/';
        }
        
        // Safe DOM manipulation - XSS protected
        const icon = document.createElement('i');
        icon.className = 'fas fa-archive me-2';
        icon.setAttribute('aria-hidden', 'true');
        
        link.appendChild(icon);
        link.appendChild(document.createTextNode('Archive'));
        
        // Add descriptive text for screen readers
        link.setAttribute('aria-label', 'View all versions in archive');
        
        return link;
    }
    
    // Select the dropdown menu and add the archive link
    const dropdownMenus = document.getElementsByClassName('dropdown-menu');
    for (let i = 0; i < dropdownMenus.length; i++) {
        const archiveLink = createArchiveLink();
        dropdownMenus[i].appendChild(archiveLink);
    }
});