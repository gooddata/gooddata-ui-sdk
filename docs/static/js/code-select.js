(function ($) {
    $(document).ready(function () {
        $(".gd-docs-code-select__tab").on("click", setActive);
        $(".gd-docs-code-select__osdetection .gd-docs-code-select__tab").each(osDetection);
    });

    var setActive = function() {
        if(!$(this).hasClass("active")) {
            $(this).siblings().removeClass("active");
            $(this)
                .addClass("active")
                .parent()
                .parent()
                .find(".gd-docs-code-select__code")
                .removeClass("active")
                .filter(".lang-" + $(this).data("lang"))
                .addClass("active");
        }
    }

    var osDetection = function() {
        if(window.navigator.userAgent.indexOf($(this).data("user-agent")) >= 0) {
            $(this).trigger("click");
        }
    }

}(jQuery));