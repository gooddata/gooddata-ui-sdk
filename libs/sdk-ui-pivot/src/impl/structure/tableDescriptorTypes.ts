// (C) 2007-2022 GoodData Corporation
import { IAttributeDescriptor, IResultAttributeHeader, ITotalDescriptor } from "@gooddata/sdk-model";
import { DataSeriesDescriptor, DataSeriesId } from "@gooddata/sdk-ui";
import { ColDef, ColGroupDef, Column } from "@ag-grid-community/all-modules";

export const ColumnGroupingDescriptorId = "root";

/**
 * Recognized col types. See the respective interfaces for addition information.
 *
 * @remarks see {@link SliceCol}
 * @remarks see {@link SeriesCol}
 * @remarks see {@link ScopeCol}
 * @remarks see {@link RootCol}
 */
export type TableColType =
    | "sliceCol"
    | "sliceMeasureCol"
    | "mixedValuesCol"
    | "mixedHeadersCol"
    | "seriesCol"
    | "scopeCol"
    | "rootCol";

/**
 * Base interface for all col types.
 */
export interface TableCol {
    readonly type: TableColType;
    readonly id: string;
}

/**
 * Represents table column which displays data slice attribute element names. PivotTable interface also calls
 * these `attribute column`.
 *
 * In other words, in a table that is defined as (Row: A1, Measures: M1) and looking like:
 *
 * A1    | M1
 * ----------
 * val1  | 1
 * val2  | 2
 *
 * There is one SliceCol - for row attribute A1. In this table, there will be one row for each
 * element of attribute A1.
 *
 */
export interface SliceCol extends TableCol {
    readonly type: "sliceCol";

    /**
     * Descriptor of attribute whose elements will be in this column.
     */
    readonly attributeDescriptor: IAttributeDescriptor;

    /**
     * Descriptors of totals that are defined on this column's granularity. Note that while the total
     * descriptors are also part of the attributeDescriptor (above) the total descriptors there are a snapshot done
     * at the time of table descriptor construction.
     *
     * User may change effective totals by configuring aggregations using the table's aggregation menu. If that
     * happens the table will drive a new execution with updated totals. After it finishes, the effective totals
     * in this field will be updated.
     */
    effectiveTotals: ITotalDescriptor[];

    /**
     * Column index among all slice columns
     */
    index: number;

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    readonly fullIndexPathToHere: number[];
}

export function isSliceCol(obj: unknown): obj is SliceCol {
    return (obj as SliceCol)?.type === "sliceCol";
}

/**
 * Represents table column which displays names of metrics when transposed to rows.
 *
 * In other words, in a table that is defined as (Row: A1, Measures: M1 transposed to rows) and looking like:
 *
 *          x
 * A1    |    |
 * ----------------
 * val1  | M1 | 1
 * val2  | M1 | 2
 *
 * There is one SliceMeasureCol - for name of M1. In this table, there will be one row for each
 * element of attribute A1.
 *
 */
export interface SliceMeasureCol extends TableCol {
    readonly type: "sliceMeasureCol";

    /**
     * Column index among all slice columns
     */
    index: number;

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    readonly fullIndexPathToHere: number[];
}

export interface MixedHeadersCol extends TableCol {
    readonly type: "mixedHeadersCol";

    /**
     * Column index among all slice columns
     */
    index: number; // probably will always be 0 as there will be always just one

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    readonly fullIndexPathToHere: number[];
}
// TODO pivot operates with slices and series. Better name should be used than the introduction of attr/measures.

export interface MixedValuesCol extends TableCol {
    readonly type: "mixedValuesCol";

    /**
     * Column index among all slice columns
     */
    index: number;

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    readonly fullIndexPathToHere: number[];
}

export function isSliceMeasureCol(obj: unknown): obj is SliceMeasureCol {
    return (obj as SliceMeasureCol)?.type === "sliceMeasureCol";
}

export function isMixedHeadersCol(obj: unknown): obj is MixedHeadersCol {
    return (obj as MixedHeadersCol)?.type === "mixedHeadersCol";
}

export function isMixedValuesCol(obj: unknown): obj is MixedValuesCol {
    return (obj as MixedValuesCol)?.type === "mixedValuesCol";
}

/**
 * Represents most granular column under which computed measure data for a single data series is shown.
 *
 * In other words, in a table that is defined as (Row: A1, Cols: A2, Measures: M1) and looking like:
 *
 *       -------+-----+
 *       | A2E1 | A2E2|
 * A1    | M1   | M1  | ....
 * ------------------------
 * val1  | 1    | 3   |
 * val2  | 2    | 4   |
 *
 * There are as many series cols as there are combinations of A2 attribute elements for which it is possible
 * to calculate metric M1.
 *
 * If you are familiar with the DataViewFacade's data access infrastructure, then the analogy is simple: there
 * is one series col of each data series.
 */
export interface SeriesCol extends TableCol {
    readonly type: "seriesCol";

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    fullIndexPathToHere: number[];

    /**
     * Index among all series cols.
     */
    index: number;

    readonly children: [];

    readonly seriesId: DataSeriesId;
    readonly seriesDescriptor: DataSeriesDescriptor;
}

export function isSeriesCol(obj: unknown): obj is SeriesCol {
    return (obj as SeriesCol)?.type == "seriesCol";
}

/**
 * Tables with column attributes and measures is a table with scoped data series. Such table will have its data cols
 * grouped into a tree hierarchy. The ScopeCol is composite in this tree hierarchy, whose children may be either additional
 * scope cols or finally the series cols.
 *
 * The tree hierarchy will be as deep as there are column attributes and as wide as the number of the actual scoped
 * data series.
 *
 * In other words, in a table that is defined as (Row: A1, Cols: A2, A3, Measures: M1, M2) and looking like:
 *
 *       ---------------------------
 *       |           A1E1          |
 *       -------+-----+------------+
 *       |     A2E1   |     A2E2   |
 * A1    | M1   | M2  | M1   | M2  | ....
 * ----------------------------------------
 * val1  | 1    | 3   | ....
 * val2  | 2    | 4   | ....
 *
 * There are as many series cols as there are valid combinations of A1xA2 elements for which metric values M1 and
 * M2 can be computed. These leaves are then grouped under as many trees as there are unique values of A1. In each
 * such tree there are there are further subtrees - one for each unique value of A2. And then in each subtree there
 * are the leaves - one for each measure.
 */
export interface ScopeCol extends TableCol {
    readonly type: "scopeCol";

    readonly children: DataCol[];

    /**
     * Path of indexes to follow from root, through children in order to get to this node.
     */
    fullIndexPathToHere: number[];

    /**
     * Descriptor of attribute whose value scopes this column group.
     */
    readonly attributeDescriptor: IAttributeDescriptor;

    /**
     * The actual attribute header that scopes this column group.
     */
    readonly header: IResultAttributeHeader;

    /**
     * Descriptors of all attributes on the way to this column group - excluding the descriptor of the
     * current group.
     */
    readonly descriptorsToHere: IAttributeDescriptor[];

    /**
     * Descriptors of all attribute headers on the way to this column group - excluding the header of the
     * current group.
     */
    readonly headersToHere: IResultAttributeHeader[];

    /**
     * Checks whether this column group is part of grand total.
     */
    readonly isTotal?: boolean;

    /**
     * Checks whether this column group is part of sub-total.
     */
    readonly isSubtotal?: boolean;
}

export function isScopeCol(obj: unknown): obj is ScopeCol {
    return (obj as ScopeCol)?.type === "scopeCol";
}

/**
 * Tests whether the provided col is a {@link ScopeCol} with no children - which implies that it is the bottom
 * most group that is not linked to any measure columns, which in turn means the table is not defined with any measures.
 *
 * @param col - col to test
 */
export function isEmptyScopeCol(col: AnyCol): boolean {
    return isScopeCol(col) && col.children.length === 0;
}

/**
 * Appears in a table with grouped data columns. This is the root header that has all first-level
 * group headers as children.
 */
export interface RootCol extends TableCol {
    readonly type: "rootCol";
    readonly id: "root";

    readonly children: DataCol[];
    readonly groupingAttributes: IAttributeDescriptor[];
    readonly fullIndexPathToHere: [0];
}

export function isRootCol(obj: unknown): obj is RootCol {
    return (obj as RootCol)?.type === "rootCol";
}

/**
 * Leaf data cols appear as leaf / bottom-most columns in the data part of the table. This is typically a series col,
 * however in cases of table without measures it can also be a scope col.
 */
export type LeafDataCol = SeriesCol | ScopeCol;

/**
 * Data col is a composite structure describing the data part of the table.
 */
export type DataCol = RootCol | LeafDataCol;

/**
 * All types of slice cols.
 */
export type AnySliceCol = SliceCol | SliceMeasureCol;

/**
 * Any table col. May be either the col describing the table slicing or col describing the data part of the table.
 */
export type AnyCol = SliceCol | SliceMeasureCol | MixedHeadersCol | MixedValuesCol | DataCol | ScopeCol;

/**
 * Descriptors of all table columns. The table columns are divided into two groups:
 *
 * -  Slice Columns; these describe the columns that contain description of each row of data. They are
 *    constructed from data slicing attributes.
 *
 * -  Data columns; these describe the columns that contain the actual data points. They are constructed from
 *    the data series identified in the result. A data series represents a stream of data points computed for particular
 *    measure. The data points can be sliced by some attributes and/or scoped by some attributes. Read more about
 *    that here: {@link IDataSeries}.
 *
 *    When the data series use scoping, the data cols are grouped by the scoping attribute
 *    element values and form a tree hierarchy. The hierarchy always starts with a RootCol whose children
 *    are all scope cols created from the first scoping attribute. These in turn have children that are scope
 *    cols created from the second scoping attribute. This goes on until the last layer of scoping cols whose
 *    children will be the actual SeriesCols.
 *
 *    Note 1: if the table does contain only scoping attributes but  no measures, then the last layer of scoping
 *    cols will have no children.
 *
 *    Note 2: if the data series are not scoped, then no grouping is needed and the root columns and leaf columns will
 *    contain the same values.
 *
 */
export type TableCols = {
    /**
     * All table row headers - these are derived from slicing attributes
     */
    readonly sliceCols: SliceCol[];

    /**
     * Column used for metric names when table is transposed (metrics are in rows)
     */
    readonly sliceMeasureCols: SliceMeasureCol[];

    readonly mixedHeadersCols: MixedHeadersCol[];

    /**
     * Column with values of either elements or metrics or both mixed together
     */
    readonly mixedValuesCols: MixedValuesCol[];

    /**
     * Root table cols.
     */
    readonly rootDataCols: DataCol[];

    /**
     * Leaf data cols. These are either all series columns or in a corner case all scoping cols. This latter
     * can happen when the table is constructed with scoping attributes but without measures.
     */
    readonly leafDataCols: Array<LeafDataCol>;

    /**
     * Mapping between header ID to header descriptor (useful for lookups as ColDefs 'field' references this)
     */
    readonly idToDescriptor: Record<string, AnyCol>;

    /**
     * Descriptors for attributes that are used to construct the scoping cols.
     */
    readonly scopingAttributes: IAttributeDescriptor[];
};

/**
 * Ag-Grid colDefs constructed to match the table header descriptors. See {@link TableCols} for more
 * information about nomenclature.
 */
export type TableColDefs = {
    readonly sliceColDefs: ReadonlyArray<ColDef>;
    readonly rootDataColDefs: ReadonlyArray<ColDef | ColGroupDef>;
    readonly leafDataColDefs: ReadonlyArray<ColDef>;
    readonly idToColDef: Record<string, ColDef | ColGroupDef>;
};

function isColDef(obj: unknown): obj is ColDef {
    return (obj as ColDef)?.colId !== undefined;
}

function isColumn(obj: unknown): obj is Column {
    return (obj as Column)?.getColDef !== undefined;
}

/**
 * Convenience accessor of ag-grid column identifier. This can take either string itself or the
 * different type of col flying around ag-grid API (ColDef, ColGroupDef or Column).
 *
 * For purposes of our table, column ID for ag-grid groups is the groupId.
 *
 * @param colOrId - column id (returned as is) or one of the ag-grid's column types
 */
export function agColId(colOrId: string | ColDef | ColGroupDef | Column): string {
    if (typeof colOrId === "string") {
        return colOrId;
    }
    if (isColDef(colOrId)) {
        return colOrId.colId!;
    } else if (isColumn(colOrId)) {
        return agColId(colOrId.getColDef());
    } else {
        return colOrId.groupId!;
    }
}
