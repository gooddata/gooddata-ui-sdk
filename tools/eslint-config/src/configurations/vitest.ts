// (C) 2025 GoodData Corporation

import type { IConfiguration } from "../types.js";

const configuration: IConfiguration<"@vitest"> = {
    packages: [
        {
            name: "@vitest/eslint-plugin",
            version: "1.3.5",
        },
    ],
    plugin: "@vitest",
    extends: ["plugin:@vitest/legacy-recommended"],
    rules: {
        "@vitest/expect-expect": "off",
        "@vitest/no-commented-out-tests": "warn",
        "@vitest/valid-title": "warn",
        "@vitest/no-disabled-tests": "warn",
        "@vitest/no-focused-tests": "warn",
        "@vitest/no-identical-title": "warn",
        "@vitest/valid-expect": "warn",
    },
};

export default configuration;
