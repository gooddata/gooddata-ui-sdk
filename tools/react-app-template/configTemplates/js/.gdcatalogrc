{
  "hostname": "https://public-examples.gooddata.com",
  "workspaceId": "demo",
  "output": "src/catalog.ts",
  "backend": "tiger",
  "demo": true
}