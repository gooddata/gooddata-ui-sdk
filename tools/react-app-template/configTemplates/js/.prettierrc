{
    "printWidth": 110,
    "tabWidth": 4,
    "trailingComma": "all"
}
