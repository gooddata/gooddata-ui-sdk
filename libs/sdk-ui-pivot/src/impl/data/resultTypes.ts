// (C) 2007-2022 GoodData Corporation
import { IMappingHeader } from "@gooddata/sdk-ui";
import { ROW_TOTAL } from "../base/constants.js";
import { IMeasureDescriptor, IAttributeDescriptor } from "@gooddata/sdk-model";

/**
 * All non-grand-total rows in the grid conform to this interface.
 *
 * CAREFUL: if you are updating this type, adding new props etc, look at stickyRowHandler code,
 * the updateStickyRowContentClassesAndData function - there may be a bad interplay that you would not expect.
 *
 * TODO: we need to refactor this type. wildly mixing a number of our custom props with the values for fields
 *  leads to unnecessary dangers.
 */
export interface IGridRow {
    /**
     * Mapping of column ID ⇒ data value
     */
    [key: string]: any;

    /**
     * Mapping of slice column ID ⇒ mapping header (detail about slice value)
     */
    headerItemMap: {
        [key: string]: IMappingHeader;
    };

    /**
     * If metrics are in rows each row keeps corresponding metric descriptor
     */
    measureDescriptor?: IMeasureDescriptor;

    /**
     * If column attributes header position is set to left, keeps corresponding attribute descriptor for each row.
     */
    columnAttributeDescriptor?: IAttributeDescriptor;

    /**
     * If this is 'special' row such as total or subtotal, then the 'type' will be set to value of
     * ROW_TOTAL or ROW_SUBTOTAL constant. Otherwise the field is not present.
     */
    type?: string;

    subtotalStyle?: string;
}

/**
 * Grand total rows in the table conform to this
 */
export interface IGridTotalsRow {
    type: string;
    colSpan: {
        count: number;
        headerKey: string;
    };
    calculatedForColumns?: string[];

    [key: string]: any;
}

export interface IAgGridPage {
    rowData: IGridRow[];
    rowTotals: IGridTotalsRow[];
}

export function isGridTotalsRow(obj: unknown): obj is IGridTotalsRow {
    return (obj as IGridTotalsRow)?.type === ROW_TOTAL;
}
