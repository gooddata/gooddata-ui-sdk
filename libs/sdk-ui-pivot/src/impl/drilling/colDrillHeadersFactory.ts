// (C) 2007-2025 GoodData Corporation
import {
    AnyCol,
    ScopeCol,
    SeriesCol,
    isScopeCol,
    isSeriesCol,
    isSliceCol,
    SliceCol,
    isMixedValuesCol,
    MixedValuesCol,
} from "../structure/tableDescriptorTypes.js";
import { IMappingHeader } from "@gooddata/sdk-ui";
import { IAttributeDescriptor, isResultAttributeHeader } from "@gooddata/sdk-model";
import { invariant, InvariantError } from "ts-invariant";
import { IGridRow } from "../data/resultTypes.js";
import { ColumnHeadersPosition } from "../../publicTypes.js";

export function createDataColLeafHeaders(col: SeriesCol): IMappingHeader[] {
    const mappingHeaders: IMappingHeader[] = [];

    if (col.seriesDescriptor.attributeDescriptors) {
        col.seriesDescriptor.attributeDescriptors.forEach(
            (attributeDescriptor: IAttributeDescriptor, index: number) => {
                const attributeElementDescriptor = col.seriesDescriptor.attributeHeaders![index];
                mappingHeaders.push(attributeElementDescriptor);
                mappingHeaders.push(attributeDescriptor);
            },
        );
    }
    mappingHeaders.push(col.seriesDescriptor.measureDescriptor);

    return mappingHeaders;
}

/**
 * For transposed table add metric descriptor from row
 */
export function createScopeColWithMetricHeaders(
    col: ScopeCol,
    includeColGroupHeaders: boolean,
    row?: IGridRow,
): IMappingHeader[] {
    const mappingHeaders: IMappingHeader[] = includeColGroupHeaders ? createDataColGroupHeaders(col) : [];

    if (row?.measureDescriptor) {
        mappingHeaders.push(row.measureDescriptor);
    }

    return mappingHeaders;
}

export function createMixedValuesColHeaders(col: MixedValuesCol, row: IGridRow): IMappingHeader[] {
    const mappingHeaders: IMappingHeader[] = [];

    if (row.measureDescriptor) {
        mappingHeaders.push(row.measureDescriptor);
    }

    if (row.columnAttributeDescriptor) {
        const attributeElement = row.headerItemMap[col.id];

        mappingHeaders.push(attributeElement);
        mappingHeaders.push(row.columnAttributeDescriptor);
    }

    return mappingHeaders;
}

export function createSliceColHeaders(col: SliceCol, row: IGridRow): IMappingHeader[] {
    const result: IMappingHeader[] = [];

    if (!row.headerItemMap) {
        return [];
    }

    const attributeElement = row.headerItemMap[col.id];
    if (!attributeElement) {
        return [];
    }

    invariant(isResultAttributeHeader(attributeElement), `bad header for row data ${col.id}`);

    result.push(attributeElement);
    result.push(col.attributeDescriptor);

    return result;
}

function createDataColGroupHeaders(col: ScopeCol): IMappingHeader[] {
    const mappingHeaders: IMappingHeader[] = [];

    col.descriptorsToHere.forEach((descriptor, index) => {
        mappingHeaders.push(col.headersToHere[index]);
        mappingHeaders.push(descriptor);
    });

    mappingHeaders.push(col.header);
    mappingHeaders.push(col.attributeDescriptor);

    return mappingHeaders;
}

/**
 * Given column and optionally a grid row, this function will collect all descriptors and headers
 * to create 'mapping headers' that can be used for matching the column against drill predicates.
 *
 * Note: the row is optional and will only be needed when constructing headers for slicing columns (e.g. columns
 * that contain slicing attribute elements) - as the attribute element headers are in the row data.
 *
 * @param col - column to get mapping headers for
 * @param row - row
 */
export function createDrillHeaders(
    col: AnyCol,
    row: IGridRow,
    columnHeadersPosition: ColumnHeadersPosition,
    isTransposed: boolean,
): IMappingHeader[] {
    if (isSeriesCol(col)) {
        return createDataColLeafHeaders(col);
    } else if (isScopeCol(col)) {
        const includeColGroupHeaders = columnHeadersPosition === "top" && !isTransposed;
        return createScopeColWithMetricHeaders(col, includeColGroupHeaders, row);
    } else if (isSliceCol(col)) {
        // if this bombs, then the client is not calling the function at the right time. in order
        // to construct drilling headers for a slice col, both the column & the row data must be
        // available because the attribute element (essential part) is only available in the data itself
        invariant(row);

        return createSliceColHeaders(col, row);
    } else if (isMixedValuesCol(col)) {
        // if this bombs, then the client is not calling the function at the right time. in order
        // to construct drilling headers for a metrics in rows, both the column & the row data must be
        // available because the metric descriptor is only available in the data itself
        invariant(row);
        return createMixedValuesColHeaders(col, row);
    }

    throw new InvariantError(`unable to obtain drill headers for column of type ${col.type}`);
}
