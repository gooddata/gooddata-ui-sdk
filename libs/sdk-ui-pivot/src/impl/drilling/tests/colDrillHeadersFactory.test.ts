// (C) 2007-2021 GoodData Corporation

import { AnyCol } from "../../structure/tableDescriptorTypes.js";
import { IGridRow } from "../../data/resultTypes.js";
import { createTestRow, TwoMeasuresWithTwoRowAndTwoColumnAttributesDescriptor } from "./drilling.fixture.js";
import { ReferenceData } from "@gooddata/reference-workspace";
import { createDrillHeaders } from "../colDrillHeadersFactory.js";
import { describe, it, expect } from "vitest";

describe("createDrillHeaders", () => {
    const TestTable = TwoMeasuresWithTwoRowAndTwoColumnAttributesDescriptor;
    const FirstSliceCol = TestTable.headers.sliceCols[0];
    const SecondSliceCol = TestTable.headers.sliceCols[1];
    const LeafCol = TestTable.headers.leafDataCols[0];
    const FirstLevelGroupCol = TestTable.headers.rootDataCols[0].children[0];
    const SecondLevelGroupCol = TestTable.headers.rootDataCols[0].children[0].children[0];

    const Scenarios: Array<[string, AnyCol, IGridRow | undefined]> = [
        [
            "first attribute's header and descriptor from first slice column",
            FirstSliceCol,
            createTestRow(TestTable, [
                ReferenceData.ProductName.CompuSci,
                ReferenceData.Department.DirectSales,
            ]),
        ],
        [
            "second attribute's header and descriptor from second column",
            SecondSliceCol,
            createTestRow(TestTable, [
                ReferenceData.ProductName.CompuSci,
                ReferenceData.Department.DirectSales,
            ]),
        ],
        [
            "all scoping attribute's headers and descriptors and measure from measure's data col",
            LeafCol,
            undefined,
        ],
        [
            "just first level group's attribute header and descriptor from first level group",
            FirstLevelGroupCol,
            undefined,
        ],
        [
            "both parent's and second level's group attribute and descriptor from second level group",
            SecondLevelGroupCol,
            undefined,
        ],
    ];

    it.each(Scenarios)("should generate drill item containing %s", (_desc, col, row) => {
        expect(createDrillHeaders(col, row, "top", false)).toMatchSnapshot();
    });
});
