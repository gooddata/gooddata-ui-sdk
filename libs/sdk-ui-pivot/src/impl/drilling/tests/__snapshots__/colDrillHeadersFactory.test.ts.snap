// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createDrillHeaders > should generate drill item containing all scoping attribute's headers and descriptors and measure from measure's data col 1`] = `
[
  {
    "attributeHeaderItem": {
      "name": "Exclude",
      "uri": "Exclude",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_opportunitysnapshot.forecastcategory_id",
        "name": "Forecast Category",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_opportunitysnapshot.forecastcategory_id",
      "localIdentifier": "a_f_opportunitysnapshot.forecastcategory_id",
      "name": "Forecast Category",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
  {
    "attributeHeaderItem": {
      "name": "East Coast",
      "uri": "East Coast",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_owner.region_id",
        "name": "Region",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_owner.region_id",
      "localIdentifier": "a_f_owner.region_id",
      "name": "Region",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
  {
    "measureHeaderItem": {
      "format": "$#,##0.00",
      "identifier": "87a053b0-3947-49f3-b0c5-de53fd01f050",
      "localIdentifier": "m_87a053b0_3947_49f3_b0c5_de53fd01f050",
      "name": "Amount",
      "ref": undefined,
    },
  },
]
`;

exports[`createDrillHeaders > should generate drill item containing both parent's and second level's group attribute and descriptor from second level group 1`] = `
[
  {
    "attributeHeaderItem": {
      "name": "Exclude",
      "uri": "Exclude",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_opportunitysnapshot.forecastcategory_id",
        "name": "Forecast Category",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_opportunitysnapshot.forecastcategory_id",
      "localIdentifier": "a_f_opportunitysnapshot.forecastcategory_id",
      "name": "Forecast Category",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
  {
    "attributeHeaderItem": {
      "name": "East Coast",
      "uri": "East Coast",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_owner.region_id",
        "name": "Region",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_owner.region_id",
      "localIdentifier": "a_f_owner.region_id",
      "name": "Region",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
]
`;

exports[`createDrillHeaders > should generate drill item containing first attribute's header and descriptor from first slice column 1`] = `
[
  {
    "attributeHeaderItem": {
      "name": "CompuSci",
      "uri": "CompuSci",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "attr.f_product.product",
        "name": "Product",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "label.f_product.product.name",
      "localIdentifier": "a_label.f_product.product.name",
      "name": "Product Name",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
]
`;

exports[`createDrillHeaders > should generate drill item containing just first level group's attribute header and descriptor from first level group 1`] = `
[
  {
    "attributeHeaderItem": {
      "name": "Exclude",
      "uri": "Exclude",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_opportunitysnapshot.forecastcategory_id",
        "name": "Forecast Category",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_opportunitysnapshot.forecastcategory_id",
      "localIdentifier": "a_f_opportunitysnapshot.forecastcategory_id",
      "name": "Forecast Category",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
]
`;

exports[`createDrillHeaders > should generate drill item containing second attribute's header and descriptor from second column 1`] = `
[
  {
    "attributeHeaderItem": {
      "name": "Direct Sales",
      "uri": "Direct Sales",
    },
  },
  {
    "attributeHeader": {
      "formOf": {
        "identifier": "f_owner.department_id",
        "name": "Department",
        "ref": undefined,
        "uri": "",
      },
      "granularity": null,
      "identifier": "f_owner.department_id",
      "localIdentifier": "a_f_owner.department_id",
      "name": "Department",
      "ref": undefined,
      "totalItems": [],
      "uri": "",
    },
  },
]
`;
